src/Extensions/FakeVimProxy.cpp

```diff
@@ -0,0 +1,506 @@
+        }
+        block = block.next();
+    }
+}
+
+void FakeVimProxy::checkForElectricCharacter(bool *result, QChar c)
+{
+    *result = c == '{' || c == '}';
+}
+
+int FakeVimProxy::firstNonSpace(const QString &text)
+{
+    int indent = 0;
+    while (indent < text.length() && text.at(indent) == ' ')
+        ++indent;
+    return indent;
+}
+
+void FakeVimProxy::updateExtraSelections()
+{
+    QTextEdit *editor = qobject_cast<QTextEdit *>(m_widget);
+    QPlainTextEdit *plainEditor = qobject_cast<QPlainTextEdit *>(m_widget);
+    if (editor)
+    {
+        editor->setExtraSelections(m_clearSelection + m_searchSelection + m_blockSelection);
+    }
+    else if (plainEditor)
+    {
+        plainEditor->setExtraSelections(m_clearSelection + m_searchSelection + m_blockSelection);
+    }
+}
+
+bool FakeVimProxy::wantSaveAndQuit(const FakeVim::Internal::ExCommand &cmd)
+{
+    return cmd.cmd == "wq";
+}
+
+bool FakeVimProxy::wantSave(const FakeVim::Internal::ExCommand &cmd)
+{
+    return cmd.matches("w", "write") || cmd.matches("wa", "wall");
+}
+
+bool FakeVimProxy::wantQuit(const FakeVim::Internal::ExCommand &cmd)
+{
+    return cmd.matches("q", "quit") || cmd.matches("qa", "qall");
+}
+
+bool FakeVimProxy::save()
+{
+    if (!hasChanges())
+        return true;
+    auto *window = qobject_cast<MainWindow *>(m_mainWindow);
+    if (window)
+        return window->save(true, "Vim Save");
+
+    return false;
+}
+
+void FakeVimProxy::cancel()
+{
+    auto *window = qobject_cast<MainWindow *>(m_mainWindow);
+    if (window)
+    {
+        if (hasChanges())
+        {
+            if (window->closeConfirm())
+                window->closeWindow();
+        }
+        else
+            invalidate();
+    }
+}
+void FakeVimProxy::invalidate()
+{
+    auto *window = qobject_cast<MainWindow *>(m_mainWindow);
+    if (window)
+        window->closeWindow();
+}
+
+bool FakeVimProxy::hasChanges()
+{
+    auto *window = qobject_cast<MainWindow *>(m_mainWindow);
+    if (window)
+        return window->isTextChanged();
+    return false;
+}
+
+QTextDocument *FakeVimProxy::document() const
+{
+    QTextDocument *doc = NULL;
+    if (QPlainTextEdit *ed = qobject_cast<QPlainTextEdit *>(m_widget))
+        doc = ed->document();
+    else if (QTextEdit *ed = qobject_cast<QTextEdit *>(m_widget))
+        doc = ed->document();
+    return doc;
+}
+
+QString FakeVimProxy::content() const
+{
+    return document()->toPlainText();
+}
+
+void FakeVimProxy::initHandler(FakeVim::Internal::FakeVimHandler *handler)
+{
+    handler->handleCommand(_("set nopasskeys"));
+    handler->handleCommand(_("set nopasscontrolkey"));
+
+    // Set some Vim options.
+    handler->handleCommand(_("set expandtab"));
+    handler->handleCommand(_("set shiftwidth=8"));
+    handler->handleCommand(_("set tabstop=%1").arg(SettingsHelper::getTabWidth()));
+    handler->handleCommand(_("set autoindent"));
+    handler->handleCommand(_("set smartindent"));
```

Comment:
Maybe some code editor settings can be set here? At least, now the tab width is not set properly. I'm wondering why `set shiftwidth=8`.

Reply:
Yes, that's why I want to make a separate settings page for FakeVim.

Reply:
Shift width and tab width are explained [here](https://www.reddit.com/r/vim/comments/99ylz8/confused_about_the_difference_between_tabstop_and/)

Reply:
> Yes, that's why I want to make a separate settings page for FakeVim.

I don't get it. I meant if you want, you can apply some settings in the Code Edit section to FakeVim. It can't be the reason you want to make a separate page.

Do you want to add a convenient way to configure Fake Vim besides writing a `.vimrc`?

Reply:
Initially I was thinking of it but if you say, I can reuse existing Code Edit options like Tab Width. Replace with Space (expandtabs) etc.

Reply:
Basically, what I want is, if a setting in the Code Edit section is not used by Fake Vim, it should be disabled (i.e. it depends on Fake Vim is not enabled).

Reply:
And, if there are only the "Enable" and "vimrc" settings, I think it should in the Code Edit page. At least, it's not an extension. I thought an extension is a third party software not included in CP Editor itself.

Reply:
Moved to Core and Settings moved to Code Edit